File: extensions/defenseclaw/src/scanners/plugin_scanner/llm_client.ts
/**
 * LLM client — calls litellm via the defenseclaw Python bridge.
 *
 * Uses the same litellm library as the skill scanner so provider routing,
 * API key handling, and model support are identical.
 *
 * The bridge is at cli/defenseclaw/llm.py and accepts JSON on stdin,
 * returns JSON on stdout.
 */
import { execFileSync } from "node:child_process";
import { existsSync } from "node:fs";
import { resolve } from "node:path";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface LLMConfig {
  /** Model name (e.g. "claude-sonnet-4-20250514", "gpt-4", "ollama/llama3"). */
  model: string;
  /** API key (falls back to env vars if not set). */
  apiKey?: string;
  /** Custom API base URL (for Azure, Ollama, etc.). */
  apiBase?: string;
  /** Provider hint (litellm auto-detects from model name). */
  provider?: string;
  /** Max output tokens (default 8192). */
  maxTokens?: number;
  /** Number of consensus runs (default 1). */
  consensusRuns?: number;
  /** Python binary path (default "python3"). */
  pythonBinary?: string;
}

export interface LLMMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

export interface LLMResponse {
  content: string;
  model: string;
  usage: { prompt_tokens?: number; completion_tokens?: number; total_tokens?: number };
  error: string | null;
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

const ALLOWED_PYTHON_NAMES = new Set(["python3", "python", "python3.11", "python3.12", "python3.13"]);

export function validatePythonBinary(raw: string): string {
  if (ALLOWED_PYTHON_NAMES.has(raw)) {
    return raw;
  }

  const resolved = resolve(raw);
  if (
    !resolved.startsWith("/") ||
    resolved.includes("..") ||
    !existsSync(resolved)
  ) {
    throw new Error(
      `Refusing untrusted python_binary: "${raw}". ` +
      `Use an absolute path to an existing executable or one of: ${[...ALLOWED_PYTHON_NAMES].join(", ")}`,
    );
  }
  return resolved;
}

export async function callLLM(
  config: LLMConfig,
  messages: LLMMessage[],
): Promise<LLMResponse> {
  const python = validatePythonBinary(config.pythonBinary ?? "python3");

  const request = {
    model: config.model,
    messages,
    max_tokens: config.maxTokens ?? 8192,
    temperature: 0.0,
    ...(config.apiKey && { api_key: config.apiKey }),
    ...(config.apiBase && { api_base: config.apiBase }),
    ...(config.provider && { provider: config.provider }),
  };

  const input = JSON.stringify(request);

  try {
    const stdout = execFileSync(
      python,
      ["-m", "defenseclaw.llm"],
      { input, timeout: 120_000, maxBuffer: 10 * 1024 * 1024, encoding: "utf-8" },
    );

    const response = JSON.parse(stdout) as LLMResponse;
    return response;
  } catch (err: unknown) {
    const message = err instanceof Error ? err.message : String(err);
    return { content: "", model: config.model, usage: {}, error: message };
  }
}

/**
 * Run LLM N times and return only findings that appear in the majority of runs.
 * This matches the skill scanner's consensus_runs feature.
 */
export async function callLLMWithConsensus(
  config: LLMConfig,
  messages: LLMMessage[],
  runs: number,
): Promise<LLMResponse> {
  if (runs <= 1) return callLLM(config, messages);

  const results = await Promise.all(
    Array.from({ length: runs }, () => callLLM(config, messages)),
  );

  // Filter out errors
  const successful = results.filter((r) => !r.error);
  if (successful.length === 0) {
    return results[0]; // return first error
  }

  // For consensus, return the response with the most overlap in content.
  // Simple heuristic: pick the longest non-error response (it likely
  // contains the most complete analysis).
  const best = successful.reduce((a, b) =>
    a.content.length >= b.content.length ? a : b,
  );

  // Sum usage across runs
  const totalUsage = {
    prompt_tokens: successful.reduce((s, r) => s + (r.usage.prompt_tokens ?? 0), 0),
    completion_tokens: successful.reduce((s, r) => s + (r.usage.completion_tokens ?? 0), 0),
    total_tokens: successful.reduce((s, r) => s + (r.usage.total_tokens ?? 0), 0),
  };

  return { ...best, usage: totalUsage };
}

File: cli/defenseclaw/scanner/plugin_scanner/llm_client.py
"""LLM client -- calls litellm via the defenseclaw bridge."""

from __future__ import annotations

import json
import os
import subprocess
from dataclasses import dataclass, field
from typing import Any


@dataclass
class LLMConfig:
    model: str = ""
    api_key: str | None = None
    api_base: str | None = None
    provider: str | None = None
    max_tokens: int | None = None
    python_binary: str | None = None


@dataclass
class LLMMessage:
    role: str  # "system" | "user" | "assistant"
    content: str


@dataclass
class LLMResponse:
    content: str = ""
    model: str = ""
    usage: dict[str, int] = field(default_factory=dict)
    error: str | None = None


# ---------------------------------------------------------------------------
# Client
# ---------------------------------------------------------------------------

_ALLOWED_PYTHON_NAMES = {"python3", "python", "python3.11", "python3.12", "python3.13"}


def validate_python_binary(raw: str) -> str:
    if raw in _ALLOWED_PYTHON_NAMES:
        return raw

    resolved = os.path.abspath(raw)
    if not os.path.isabs(resolved) or ".." in resolved or not os.path.exists(resolved):
        allowed = ", ".join(sorted(_ALLOWED_PYTHON_NAMES))
        raise ValueError(
            f'Refusing untrusted python_binary: "{raw}". '
            f"Use an absolute path to an existing executable or one of: {allowed}"
        )
    return resolved


def call_llm(
    config: LLMConfig | dict[str, Any],
    messages: list[LLMMessage] | list[dict[str, str]],
) -> LLMResponse:
    # Normalise config
    if isinstance(config, dict):
        model = config.get("model", "")
        api_key = config.get("api_key")
        api_base = config.get("api_base")
        provider = config.get("provider")
        max_tokens = config.get("max_tokens", 8192)
        python_binary = config.get("python_binary", "python3")
    else:
        model = config.model
        api_key = config.api_key
        api_base = config.api_base
        provider = config.provider
        max_tokens = config.max_tokens or 8192
        python_binary = config.python_binary or "python3"

    python = validate_python_binary(python_binary or "python3")

    # Normalise messages
    msg_dicts = []
    for m in messages:
        if isinstance(m, dict):
            msg_dicts.append(m)
        else:
            msg_dicts.append({"role": m.role, "content": m.content})

    request: dict[str, Any] = {
        "model": model,
        "messages": msg_dicts,
        "max_tokens": max_tokens,
        "temperature": 0.0,
    }
    if api_key:
        request["api_key"] = api_key
    if api_base:
        request["api_base"] = api_base
    if provider:
        request["provider"] = provider

    input_json = json.dumps(request)

    try:
        proc = subprocess.run(
            [python, "-m", "defenseclaw.llm"],
            input=input_json,
            capture_output=True,
            text=True,
            timeout=120,
        )
        response = json.loads(proc.stdout)
        return LLMResponse(
            content=response.get("content", ""),
            model=response.get("model", model),
            usage=response.get("usage", {}),
            error=response.get("error"),
        )
    except Exception as e:
        return LLMResponse(content="", model=model, usage={}, error=str(e))
